src/libsyntax/parse/parser.rs

```diff
@@ -5232,23 +5234,32 @@ impl<'a> Parser<'a> {
             }
             let mut err = self.diagnostic().struct_span_err(id_sp,
                 "cannot declare a new module at this location");
-            let this_module = match self.directory.path.file_name() {
-                Some(file_name) => file_name.to_str().unwrap().to_owned(),
-                None => self.root_module_name.as_ref().unwrap().clone(),
-            };
-            err.span_note(id_sp,
-                          &format!("maybe move this module `{0}` to its own directory \
-                                     via `{0}/mod.rs`",
-                                    this_module));
+            if id_sp != syntax_pos::DUMMY_SP {
+                let mut src_path = PathBuf::from(self.sess.codemap().span_to_filename(id_sp));
+                let src_path_copy = src_path.clone();
+                let stem = src_path_copy.file_stem().unwrap_or(OsStr::new(""));
+                let mut dest_path = src_path.clone();
+                dest_path.set_file_name(stem);
+                dest_path.push("mod.rs");
+                src_path.set_extension("rs");
```

Comment:
Why do this? If the source path isn't `.rs`, we'll be telling the user to move a file that doesn't exist.